examples/whl.py

```diff
@@ -0,0 +1,120 @@
+"""Install from a .whl file.
+"""
+
+import argparse
+import codecs
+import contextlib
+import sys
+import zipfile
+
+from installer._compat import pathlib
+from installer._compat.typing import TYPE_CHECKING
+from installer.layouts import DistInfo
+from installer.records import RecordItem, parse_record_file, write_record_file
+
+if TYPE_CHECKING:
+    from typing import Any, ContextManager, Dict, IO, Iterator
+
+
+class ZipFileInstaller(object):
+    """Install a local wheel.
+    """
+
+    def __init__(self, name, distinfo, zf):
+        # type: (str, DistInfo, zipfile.ZipFile) -> None
+        self._name = name
+        self._distinfo = distinfo
+        self._zf = zf
+
+    @classmethod
+    @contextlib.contextmanager
+    def create(cls, name, wheel_path):
+        # type: (str, pathlib.Path) -> Iterator[ZipFileInstaller]
+        project_name, project_version, _ = wheel_path.stem.split("-", 2)
+        with zipfile.ZipFile(str(wheel_path)) as zf:
+            entry_names = (name.lstrip("/").split("/", 1)[0] for name in zf.namelist())
+            distinfo = DistInfo.find(project_name, project_version, entry_names)
+            yield cls(name, distinfo, zf)
+
+    @contextlib.contextmanager
+    def _open_adjacent_tmp_for_write(self, path, **kwargs):
+        # type: (pathlib.Path, Any) -> Iterator[IO[Any]]
+        path.parent.mkdir(parents=True, exist_ok=True)
+        temp = path.with_name("{}.tmp.{}".format(path.name, self._name))
+        with temp.open(**kwargs) as f:
+            yield f
+        temp.replace(path)
```

Comment:
given it also replaces I feel like just open is not adequate naming here

Reply:
then again what's the point of this tmp file rather than just direct-write?

Reply:
I stole this pattern from pip and thought the original implementation must exists for a reason. Likely some kind of edge case about writing to an already-open file.

I’m not really in love with the naming of this either, but can’t think of a better one.

Reply:
@pradyunsg care to enlighten us? for me at the moment seems just a needless complication 😊 At least let's document it why we need this awkward way.

Reply:
I tried to trace this, and… it takes me straight back to the initial implementation of wheel support! cc @dholth

https://github.com/pypa/pip/commit/aa5b33dbf4907920c3d6cec42f1385b08096ed07#diff-2695f32c4432acd141c3dbe7e7e3a6b0R152

Reply:
It's likely to prevent concurrency issues, the rename operation being atomic.

Reply:
If so I consider this a bad approach, it silently swallows an important error. Sure you will not throw a failure... but instead, you're getting an environment that might not be what you expect. I'd rather throw the failure to notify the user. E.g. imagine two processes install in parallel different versions; with this approach, we'll not throw an error, but the created version will be a combination of version A and B. I'd rather throw and let the user know that something bad is happening.

Reply:
Not to mention that while you might not get failure on the ``.dist-info`` file, you can still get the same concurrency failure on ``.dist-info.tmp`` file.